lib/_tls_wrap.js

```diff
@@ -37,211 +37,243 @@ const tls_wrap = process.binding('tls_wrap');
+  owner._SNICallback(servername, (err, context) => {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return requestOCSP(new errors.Error('ERR_MULTIPLE_CALLBACK'), owner);
     once = true;
 
     if (err)
-      return cb(err);
+      return requestOCSP(err, owner);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (owner._handle === null)
+      return requestOCSP(new errors.Error('ERR_SOCKET_CLOSED'), owner);
 
     // TODO(indutny): eventually disallow raw `SecureContext`
     if (context)
-      self._handle.sni_context = context.context || context;
+      owner._handle.sni_context = context.context || context;
 
-    cb(null, self._handle.sni_context);
+    requestOCSP(null, owner, info);
   });
 }
 
 
-function requestOCSP(self, hello, ctx, cb) {
+function requestOCSP(err, self, hello) {
+  if (err !== null)
+    return self.destroy(err);
+
   if (!hello.OCSPRequest || !self.server)
-    return cb(null);
+    return requestOCSPDone(null, self);
+
+  let ctx = self._handle.sni_context;
 
   if (!ctx)
     ctx = self.server._sharedCreds;
 
   // TLS socket is using a `net.Server` instead of a tls.TLSServer.
   // Some TLS properties like `server._sharedCreds` will not be present
   if (!ctx)
-    return cb(null);
+    return requestOCSPDone(null, self);
 
   // TODO(indutny): eventually disallow raw `SecureContext`
   if (ctx.context)
     ctx = ctx.context;
 
   if (self.server.listenerCount('OCSPRequest') === 0) {
-    return cb(null);
-  } else {
-    self.server.emit('OCSPRequest',
-                     ctx.getCertificate(),
-                     ctx.getIssuer(),
-                     onOCSP);
+    return requestOCSPDone(null, self);
   }
 
-  var once = false;
-  function onOCSP(err, response) {
+  let once = false;
+  const onOCSP = (err, response) => {
     if (once)
-      return cb(new errors.Error('ERR_MULTIPLE_CALLBACK'));
+      return requestOCSPDone(new errors.Error('ERR_MULTIPLE_CALLBACK'), self);
     once = true;
 
     if (err)
-      return cb(err);
+      return requestOCSPDone(err, self);
 
-    if (!self._handle)
-      return cb(new errors.Error('ERR_SOCKET_CLOSED'));
+    if (self._handle === null)
+      return requestOCSPDone(new errors.Error('ERR_SOCKET_CLOSED'), self);
 
     if (response)
       self._handle.setOCSPResponse(response);
-    cb(null);
-  }
-}
-
-
-function onclienthello(hello) {
-  var self = this;
-
-  loadSession(self, hello, function(err) {
-    if (err)
-      return self.destroy(err);
+    requestOCSPDone(null, self);
+  };
 
-    self._handle.endParser();
-  });
+  self.server.emit('OCSPRequest',
+                   ctx.getCertificate(),
+                   ctx.getIssuer(),
+                   onOCSP);
 }
 
+function requestOCSPDone(err, self) {
+  if (err !== null)
+    return self.destroy(err);
 
-function oncertcb(info) {
-  var self = this;
-  var servername = info.servername;
+  if (self._handle === null)
```

Comment:
If I am not mistaken, it is not possible to reach this. `self._handle` should never be null.

Reply:
Since the user has control over that callback, it could indeed be called after _handle has been nulled.